### PR TITLE
no-useless-spread // Updates fixer to handle trailing comma and empty array edge cases

### PR DESCRIPTION
**Pull request by aervin**
#### Checklist

- [X] Fixes: [#168](https://github.com/fimbullinter/wotan/issues/168)
- [X] Added or updated tests / baselines
- [] Documentation update

Went ahead and merged your working branch as well. Hope that's alright.

### CODE REVIEW

File: packages/mimir/src/rules/no-useless-spread.ts [hunk 1]
```diff
@@ -15,20 +24,51 @@ export class Rule extends AbstractRule {
...
 
-function removeUselessSpread(node: ts.SpreadElement | ts.SpreadAssignment): Replacement[] {
-    return [
+function removeUselessSpread(node: ts.SpreadElement | ts.SpreadAssignment): Replacement | Replacement[] {
+    /* Handles edge cases of spread on empty array literals */
+    if (
+        isSpreadElement(node) &&
+        isArrayLiteralExpression(node.expression) &&
+        node.expression.elements.length === 0 &&
+        node.parent &&
```
Comment (ajafff) [inline packages/mimir/src/rules/no-useless-spread.ts]: `node.parent` will always be defined at this point

File: packages/mimir/src/rules/no-useless-spread.ts [hunk 2]
```diff
@@ -15,20 +24,51 @@ export class Rule extends AbstractRule {
...
+        isArrayLiteralExpression(node.parent)
```
Comment (ajafff) [inline packages/mimir/src/rules/no-useless-spread.ts]: this should not be specific to array literals. this part of the condition should be removed to handle all possible parents of SpreadElement:

```ts
console.log('foo', ...[],);
// currently fixed to
console.log('foo',,);
// should be fixed to
console.log('foo',);

// same with constructor calls
new Foo('foo', ...[],);
```

File: packages/mimir/src/rules/no-useless-spread.ts [hunk 3]
```diff
@@ -15,20 +24,51 @@ export class Rule extends AbstractRule {
...
+    const replacements = [
         Replacement.delete(node.getStart(), node.expression.getStart() + 1),
         Replacement.delete(node.expression.end - 1, node.expression.end),
     ];
+
+    /* Handles trailing commas in array/object literals */
+    if (isSpreadElement(node) && isArrayLiteralExpression(node.expression) && node.expression.elements.hasTrailingComma)
+        replacements.push(Replacement.delete(node.expression.elements.end - 1, node.expression.elements.end));
+
+    if (
```
Comment (ajafff) [inline packages/mimir/src/rules/no-useless-spread.ts]: nit: this can be an `else if`

File: packages/mimir/src/rules/no-useless-spread.ts (hunk 2 above)
Comment (aervin) [reply packages/mimir/src/rules/no-useless-spread.ts]: Yeah you're right and the TS compiler was telling me as much... What's the cleanest or most efficient way to grab the next comma if it exists?

File: packages/mimir/src/rules/no-useless-spread.ts [hunk 4]
```diff
@@ -15,20 +24,58 @@ export class Rule extends AbstractRule {
...
+        node.expression.properties.hasTrailingComma
+    ) {
+        replacements.push(Replacement.delete(node.expression.properties.end - 1, node.expression.properties.end));
+    }
+
+    return replacements;
+}
+
+function removeUselessSpreadOfEmptyArray(node: ts.SpreadElement): Replacement {
+    const parent = node.parent!;
```
Comment (aervin) [inline packages/mimir/src/rules/no-useless-spread.ts]: Interesting. Without this statement, excessive casting becomes necessary in the cases below. Feels like a bug :man_shrugging: .

Comment (ajafff) [reply packages/mimir/src/rules/no-useless-spread.ts]: This is actually working as intended 😞. The type checker only infers types of identifiers and property access. It doesn't unwrap type assertions and non-null assertions

File: packages/mimir/src/rules/no-useless-spread.ts [hunk 5]
```diff
@@ -52,23 +46,36 @@ function removeUselessSpread(node: ts.SpreadElement | ts.SpreadAssignment): Repl
...
 function removeUselessSpreadOfEmptyArray(node: ts.SpreadElement): Replacement {
-    const parent = <ts.ArrayLiteralExpression>node.parent!;
-    const spreadElementIndex = parent.elements.findIndex((el) => el === node);
-    return spreadElementIndex === parent.elements.length - 1
-        ? Replacement.delete(node.getFullStart(), parent.elements.end)
-        : Replacement.delete(node.getFullStart(), parent.elements[spreadElementIndex + 1].pos);
+    const parent = node.parent!;
+    switch (parent.kind) {
+        case ts.SyntaxKind.ArrayLiteralExpression:
+            const elementIndex = parent.elements.findIndex((el) => el === node);
```
Comment (ajafff) [inline packages/mimir/src/rules/no-useless-spread.ts]: indexOf?

File: packages/mimir/src/rules/no-useless-spread.ts [hunk 6]
```diff
@@ -15,20 +24,58 @@ export class Rule extends AbstractRule {
...
+        isObjectLiteralExpression(node.expression) &&
+        node.expression.properties.hasTrailingComma
+    ) {
+        replacements.push(Replacement.delete(node.expression.properties.end - 1, node.expression.properties.end));
+    }
+
+    return replacements;
+}
+
+function removeUselessSpreadOfEmptyArray(node: ts.SpreadElement): Replacement {
```
Comment (ajafff) [inline packages/mimir/src/rules/no-useless-spread.ts]: argh! forgot to save my comment before submitting review...

I'd like to avoid the code duplication:

```ts
const parent = node.parent!;
const list = parent.kind === ts.SyntaxKind.ArrayLiteralExpression ? parent.elements : parent.arguments!;
const index = list.indexOf(node);
return Replacement.delete(node.pos, index === list.length - 1 ? list.end : list[index + 1].pos);
```

Something similar could also be used above to handle trailing comma

**Comment by ajafff** [comment]
@aervin Thank you for the quick fix up
